### PR TITLE
Improve read throughput with larger worker side buffer

### PR DESCRIPTION
**Pull request by bf8086**
Current Alluxio worker uses gRPC stream isReady() signal for data path flow control. gRPC has a fixed buffer size of 32KB and is not configurable(See https://github.com/grpc/proposal/pull/135). This change work around this limitation by providing our own configurable flow control(client ack). In benchmark this shows 30% improvement on domain socket read with small number of concurrent streams.

### CODE REVIEW

**Comment by bf8086** [comment]
@calvinjia @gpang PTAL.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java [hunk 1]
```diff
@@ -115,6 +116,23 @@ public void send(ReqT request, long timeoutMs) throws IOException {
...
   }
 
+  /**
+   * Sends a request. Will not wait for the stream to be ready.
+   *
+   * @param request the request
+   * @throws IOException if any error occurs
+   */
+  public void send(ReqT request) throws IOException {
+    if (mClosed || mCanceled) {
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java]: Should we check `mClosedFromRemote`?

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java [hunk 2]
```diff
@@ -141,6 +141,14 @@ public DataBuffer readChunk() throws IOException {
       return null;
     }
     mPosToRead += buffer.readableBytes();
+    if (!mStream.isClosedFromRemote()) {
+      try {
+        mStream.send(ReadRequest.newBuilder().setOffsetReceived(mPosToRead).build());
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java]: It feels like the check should be part of the send method?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 3]
```diff
@@ -2094,6 +2094,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.WORKER)
           .build();
+  public static final PropertyKey WORKER_NETWORK_READER_BUFFER_SIZE_BYTES =
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: Should we call this max bytes in flight instead?

Comment (bf8086) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: This name comes from a property we had before moving to gRPC. I also plan to reuse it once the gRPC buffer size become configurable, at which time I will remove this workaround in favor of the built-in flow control mechanism(the RPC protocol will still be backward compatible).

Comment (calvinjia) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: So this is a configuration option that was available in 1.x which did the same thing?

Comment (bf8086) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: The option in 1.x is `WORKER_NETWORK_READER_BUFFER_SIZE_PACKETS`. The flow control logic is similar except that it used receipt callback from Netty.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java (hunk 1 above)
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java]: Yeah we should. Added.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java (hunk 2 above)
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java]: Updated.

**Comment by bf8086** [comment]
@calvinjia Thanks! I updated the PR to address your comments. PTAL.

**Comment by gpang** [comment]
alluxio-bot, check this please

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java [hunk 4]
```diff
@@ -141,6 +141,12 @@ public DataBuffer readChunk() throws IOException {
       return null;
     }
     mPosToRead += buffer.readableBytes();
+    try {
+      mStream.send(ReadRequest.newBuilder().setOffsetReceived(mPosToRead).build());
+    } catch (Exception e) {
+      // nothing is done as the receipt is sent at best effort
+      LOG.warn("Failed to send receipt of data: {}.", e.getMessage());
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java]: Added.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 5]
```diff
@@ -2094,6 +2094,14 @@ public String toString() {
...
+      new Builder(Name.WORKER_NETWORK_READER_BUFFER_SIZE_BYTES)
+          .setDefaultValue("4MB")
+          .setDescription("When a client read from a remote worker, the maximum amount of data"
+              + " not received by client allowed before the worker pauses sending more data.")
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: Added.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java [hunk 6]
```diff
@@ -141,6 +141,12 @@ public DataBuffer readChunk() throws IOException {
       return null;
     }
     mPosToRead += buffer.readableBytes();
+    try {
+      mStream.send(ReadRequest.newBuilder().setOffsetReceived(mPosToRead).build());
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataReader.java]: This is a dedicated ack message sent from client for every chunk of data received from worker. The client does not sent any message to attach this ack to after the initial request is sent.

**Comment by bf8086** [comment]
Thanks @gpang! I updated the PR to address your comment. PTAL.

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Pull request merged**